kernel/kernel-api/src/main/java/io/delta/kernel/internal/replay/ActionWrapper.java

```diff
@@ -34,10 +42,44 @@ public ColumnarBatch getColumnarBatch() {
     }
 
     public boolean isFromCheckpoint() {
-        return isFromCheckpoint;
+        return checkpointPath != null;
     }
 
     public long getVersion() {
         return version;
     }
+
+    /**
+     * Reads SidecarFile actions from ColumnarBatch, appending the SidecarFile actions to the end
+     * of the file list and removing sidecar actions from the ColumnarBatch.
+     */
+    public ActionWrapper appendSidecarsFromBatch(List<FileWrapper> files) {
+        // If the source checkpoint for this action does not use sidecars, sidecars will not exist
+        // in schema.
+        CheckpointInstance sourceCheckpoint =
+                new CheckpointInstance(checkpointPath.getName());
+        if (!sourceCheckpoint.format.usesSidecars()) {
+            return this;
+        }
+
+        // Sidecars will exist in schema. Extract sidecar files, then remove sidecar files from
+        // batch output.
+        int sidecarIndex =
+                columnarBatch.getSchema().fieldNames().indexOf(LogReplay.SIDECAR_FIELD_NAME);
+        CloseableIterator<SidecarFile> sidecars =
+                columnarBatch.getRows().map(r -> SidecarFile.fromRow(r, sidecarIndex));
```

Comment:
Can `SidecarFile` take the `ColumnarBatch` and `rowIndex` and return the `SidecarFile`. The `getRows` has additional overhead with object creation.

Reply:
Fixed.